File: A2/python-skeleton/Map.py
'''
/*
* CSCI3180 Principles of Programming Languages
*
* --- Declaration ---
*
* I declare that the assignment here submitted is original except for source
* material explicitly acknowledged. I also acknowledge that I am aware of
* University policy and regulations on honesty in academic work, and of the
* disciplinary guidelines and procedures applicable to breaches of such policy
* and regulations, as contained in the website
* http://www.cuhk.edu.hk/policy/academichonesty/
*
* Assignment 2
* Name : Taneja Nikunj
* Student ID : 1155123371
* Email Addr : ntaneja9@cse.cuhk.edu.hk
*/
'''

from Cell import Cell


class Map:
    def __init__(self, height, width):
        self._rows = height
        self._cols = width
        self._cells = [[Cell() for x in range(width)] for y in range(height)]
    
    #TODO: rows getter
    @property
    def rows(self):
        return self._rows
    
    #TODO: cols getter
    @property
    def cols(self):
        return self._cols
    
    def get_cell(self, row, col):
        # TODO: check whether the position is out of boundary 
        #       if not, return the cell at (row, col)
        #       return None otherwise 

        if row < 0 or row >= self._rows or col < 0 or col >= self._cols:
            print("\033[1;31;46mNext move is out of boundary!\033[0;0m")
            return None 
        else:
            return self._cells[row][col]
        # END TODO 

    def build_cell(self, row, col, cell):
        # TODO: check whether the position is out of boundary 
        #       if not, add a cell (row, col) and return True
        #       return False otherwise 
        if row < 0 or row >= self._rows or col < 0 or col >= self._cols:
            print("\033[1;31;46mThe position (%d, %d) is out of boundary!\033[0;0m" %(row, col))
            return False 
        else:
            self._cells[row][col] = cell
            return True
        # END TODO

    # return a list of cells which are neighbours of cell (row, col) 
    def get_neighbours(self, row, col):
        return_cells = []
        # TODO: return a list of neighboring cells of cell (row, col) 
        for i in range(max(0, row-1), min(row+1, self._rows-1) + 1):
            for j in range(max(0, col-1), min(col+1, self._cols-1) + 1):
                if i == row and j == col:
                    continue
                return_cells.append(self._cells[i][j])
        return return_cells
        # END TODO
        

    def display(self):
        # TODO: print the map by calling diplay of each cell 
        print("   ", end='')
        for i in range(0, self._cols):
            print(f"{i}     ", end='')
        print()
        for i in range(0, self._rows):
            print(f"{i} ", end='')
            for j in range(0, self._cols):
                self._cells[i][j].display()
            print("\n")
        # END TODO

File: A2/java-extension/src/Map.java
import java.util.ArrayList;

public class Map {
	private int rows;
	private int cols;
	private Cell[][] cells;
	
	public Map(int rows, int cols) {
		this.rows = rows;
		this.cols = cols;
		cells = new Cell[this.rows][this.cols];
	}

	public int getCols() {
		return this.cols;
	}

	public int getRows() {
		return this.rows;
	}
	
	public Cell getCell(int row, int col) {
		if (row < 0 || row >= this.rows || col < 0 || col >= this.cols) {
			System.out.printf("\033[1;31;46mNext move is out of boundary!\033[0;0m%n");
			return null;
		} else {
			return this.cells[row][col];
		}
	}
	
	public boolean buildCell(int row, int col, Cell cell) {
		if (row >= 0 && row < this.rows && col >= 0 && col < this.cols) {
			this.cells[row][col] = cell;
			return true;
		} else {
			System.out.printf("\033[1;31;46mThe position (%d, %d) is out of boundary!\033[0;0m%n", row, col);
			return false;
		}
	}
	
	public ArrayList<Cell> getNeighbours(int row, int col) {
		ArrayList<Cell> returnCells = new ArrayList<Cell>();
		for (int i = Math.max(0, row - 1); i <= Math.min(row + 1, this.rows - 1); ++i) {
			for (int j = Math.max(0, col - 1); j <= Math.min(col + 1, this.cols - 1); ++j) {
				returnCells.add(this.cells[i][j]);
			}
		}
		return returnCells;
	}
	
	public void display() {
		System.out.printf("   ");
		for (int i = 0; i < this.cols; ++i) {
			System.out.printf("%d     ", i);
		}
		System.out.printf("%n");
		for (int i = 0; i < this.rows; ++i) {
			System.out.printf("%d ", i);
			for (int j = 0; j < this.cols; ++j) {
				this.cells[i][j].display();
			}
			System.out.printf("%n%n");
		}
	}
}